File: src/main/java/mil/nga/tiff/field/tag/FieldTagType.java
package mil.nga.tiff.field.tag;

import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

public sealed interface FieldTagType permits TiffBasicTag, TiffExtendedTag, JpegTag, ExifTag, GeoTiffTag, GdalTag, MetaTag {

    /**
     * Find all available described tags
     *
     * @return list of field tag types
     */
    static List<FieldTagType> findAll() {
        @SuppressWarnings("unchecked")
        Class<FieldTagType>[] subclasses = (Class<FieldTagType>[]) FieldTagType.class.getPermittedSubclasses();

        return Stream.of(subclasses)
            .flatMap(o -> Stream.of(o.getEnumConstants()))
            .toList();
    }

    /**
     * Get a field tag type by id
     *
     * @param id tag id
     * @return field tag type
     */
    static Optional<FieldTagType> getById(int id) {
        return findAll()
            .stream()
            .filter(o -> o.getId() == id)
            .findAny();
    }

    boolean isArray();

    int getId();
}

File: src/main/java/mil/nga/tiff/field/tag/FieldTagType.kt
package mil.nga.tiff.field.tag

interface FieldTagType {
    val isArray: Boolean
    val id: Int

    companion object {
        private val ALL: List<FieldTagType> = listOf(
                ExifTag.entries,
                GdalTag.entries,
                GeoTiffTag.entries,
                JpegTag.entries,
                MetaTag.entries,
                TiffBasicTag.entries,
                TiffExtendedTag.entries
            ).flatten();

        /**
         * Get a field tag type by id
         *
         * @param id tag id
         * @return field tag type
         */
        fun getById(id: Int): FieldTagType? {
            return ALL.firstOrNull { it.id == id }
        }
    }
}
